File: mobile-app/lib/models/log.model.dart
class LogModel {
  final String id;
  final String deviceId;
  final String action;
  final DateTime timestamp;
  final Map<String, dynamic>? metadata;

  LogModel({
    required this.id,
    required this.deviceId,
    required this.action,
    required this.timestamp,
    this.metadata,
  });

  factory LogModel.fromJson(Map<String, dynamic> json) {
    return LogModel(
      id: json['_id'] ?? '',
      deviceId: json['deviceId'] ?? '',
      action: json['action'] ?? '',
      timestamp: DateTime.parse(json['timestamp'] ?? DateTime.now().toIso8601String()),
      metadata: json['metadata'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'deviceId': deviceId,
      'action': action,
      'timestamp': timestamp.toIso8601String(),
      'metadata': metadata,
    };
  }
}

File: backend/src/models/Log.model.ts
import mongoose, { Schema, Document } from "mongoose";

export interface ILog extends Document {
  deviceId?: mongoose.Types.ObjectId;
  action:
    | "unlock"
    | "door_open"
    | "door_close"
    | "delivery_success"
    | "co_owner_added"
    | "co_owner_removed"
    | "admin_slot_update"
    | "order_requested"
    | "order_approved"
    | "order_rejected"
    | "user_signup";
  timestamp: Date;
  metadata?: Record<string, unknown>;
}

const logSchema = new Schema<ILog>(
  {
    deviceId: {
      type: Schema.Types.ObjectId,
      ref: "Device",
      required: false,
    },
    action: {
      type: String,
      required: true,
    },
    timestamp: {
      type: Date,
      default: Date.now,
    },
    metadata: {
      type: Schema.Types.Mixed,
    },
  },
  { timestamps: true }
);

logSchema.index({ deviceId: 1, timestamp: -1 });

export const Log = mongoose.model<ILog>("Log", logSchema);
